02-peer-protocol.md

```diff
@@ -50,10 +67,365 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+
+`sequence` is the sequence number of this input.  Must be less than
+4294967294 (0xFFFFFFFE). See BIP125.
+
+`script_sig` is the scriptSig for the input, with length omitted.
+The `script_sig` for non-P2SH-wrapped inputs will be an empty byte array.
+
+### The `tx_add_output` Message
+
+This message adds a transaction output.
+
+1. type: 67 (`tx_add_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u64`:`sats`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+#### Requirements
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST add all sent outputs to the transaction
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add the specified output to the transaction
+  - MUST accept P2SH, P2WSH, P2WPKH, P2PKH, P2TR `script`s
+  - MAY fail the negotiation if `script` is non-standard
+  - MUST fail the negotiation if:
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - it has received 4096 `tx_add_output` messages during this negotiation
+    - the `sats` amount is less than the `dust_limit`
+    - the `sats` amount is greater than 21,0000,0000 (`MAX_MONEY`)
+
+
+#### Rationale
+Each node must know the set of the transaction outputs.
+
+`serial_id` is a randomly chosen number which uniquely identifies this output.
+Outputs in the constructed transaction are sorted by `serial_id`.
+
+`sats` is the satoshi value of the output.
+
+`script` is the scriptPubKey for the output. The length is omitted.
+It's left undefined if you accept other standard outputs such as `OP_RETURN`
+or implement stricter checks on standardness.
+
+
+### The `tx_remove_input` and `tx_remove_output` Messages
+These message removes an input from the transaction.
+
+1. type: 68 (`tx_remove_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+
+This message removes an output from the transaction.
+
+1. type: 69 (`tx_remove_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove` with a `serial_id` it did not add
+    to the transaction or has already removed
+
+The receiving node:
+  - MUST remove the indicated input or output from the transaction
+  - MUST fail the negotiation if:
+    - the input or output identified by the `serial_id` was not added by the
+      sender
+    - the `serial_id` does not correspond to a currently added input (or output)
+
+
+### The `tx_complete` Message
+
+This message signals the conclusion of a peer's transaction
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+
+#### Requirements
+
+The nodes:
+  - MUST send this message in succession to conclude this protocol
+
+The receiving node:
```

Comment:
If the receiving node still want to add things it can, yes? 

Basically "once a node has both sent and received a tx_complete message for the channel_id ..." so on and so forth?

Reply:
The sent/received `tx_complete` needs to be sequential for the negotiation to be considered concluded. Consider the following case:

```
A                           B
tx_complete ->
              <-  tx_add_input
tx_add_input ->
              <- tx_complete
```

Although a `tx_complete` has been exchanged by both parties, they weren't sequential therefore the interactive tx construction protocol is considered still in force.

`A` may send another `tx_complete` at this point to end the negotiation; they may also choose to send `tx_add_output` etc and allow `B` to respond again.

Reply:
ok I'm guessing I misunderstood what "succession" meant or missed that line entirely as per the above. it's clear